activerecord/lib/active_record/connection_adapters/pool_config.rb

```diff
@@ -51,7 +51,16 @@ def disconnect!
       def pool
         ActiveSupport::ForkTracker.check!
 
-        @pool || synchronize { @pool ||= ConnectionAdapters::ConnectionPool.new(self) }
+        @pool || synchronize { @pool ||= pool_class.new(self) }
+      end
+
+      def pool_class
+        klass = db_config.configuration_hash.fetch(:pool_class, ConnectionAdapters::ConnectionPool)
```

Comment:
I don't know how I feel about this implementation. In general the class in the pool is the class that established the connection. Now we're introducing another class that you then would need to re-implement all the methods on pool to use. Is there another way to accomplish this?

Reply:
Maybe worth investigating the design of Sequel, since they have got this part of the design working really well?